src/ape/managers/project.py

```diff
@@ -4,28 +4,257 @@
+
+
+class GithubDependency(DependencyAPI):
+    """
+    A dependency from Github. Use the ``github`` key in your ``dependencies:``
+    section of your ``ape-config.yaml`` file to declare a dependency from GitHub.
+
+    Config example::
+
+        dependencies:
+          - name: OpenZeppelin
+            github: OpenZeppelin/openzeppelin-contracts
+            version: 4.4.0
+    """
+
+    github: str
+    """
+    The Github repo ID e.g. the organization name followed by the repo name,
+    such as ``dapphub/erc20``.
+    """
+
+    branch: Optional[str] = None
+    """
+    The branch to use. **NOTE**: Will be ignored if given a version.
+    """
+
+    @property
+    def version_id(self) -> str:
+        if self.branch:
+            return self.branch
+
+        if self.version and self.version != "latest":
+            return self.version
+
+        latest_release = github_client.get_release(self.github, "latest")
+        return latest_release.tag_name
+
+    def __repr__(self):
+        return f"<{self.__class__.__name__} github={self.github}>"
+
+    def extract_manifest(self) -> PackageManifest:
+        if self.cached_manifest:
+            # Already downloaded
+            return self.cached_manifest
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_project_path = (Path(temp_dir) / self.name).resolve()
+            temp_project_path.mkdir(exist_ok=True, parents=True)
+
+            if self.branch:
+                github_client.clone_repo(self.github, temp_project_path, branch=self.branch)
+
+            else:
+                github_client.download_package(
+                    self.github, self.version or "latest", temp_project_path
+                )
+
+            return self._extract_local_manifest(temp_project_path)
+
+
+class _DependencyManager:
+    DATA_FOLDER: Path
+
+    plugin_manager: ClassVar[PluginManager] = injected_before_use()  # type: ignore
+    project_manager: ClassVar["ProjectManager"] = injected_before_use()  # type: ignore
+
+    def __init__(self, data_folder: Path):
+        self.DATA_FOLDER = data_folder
+
+    @cached_property
+    def dependency_types(self) -> Dict[str, Type[DependencyAPI]]:
+        dependency_classes = {}
+
+        for _, (config_key, dependency_class) in self.plugin_manager.dependencies:
+            dependency_classes[config_key] = dependency_class
+
+        dependency_classes["github"] = GithubDependency
+        dependency_classes["local"] = LocalDependency
+        return dependency_classes
+
+    def decode_dependency(self, config_dependency_data: Dict) -> DependencyAPI:
+        for key, dependency_cls in self.dependency_types.items():
+            if key in config_dependency_data:
+                return dependency_cls(
+                    **config_dependency_data,
+                    project_manager=self.project_manager,
+                    _data_folder=self.DATA_FOLDER,
+                )  # type: ignore
```

Comment:
When pydantic refactor, look into more smartly processing this with `Union` on `ConfigItem`

Reply:
I have looked into this already struggled due to the fact that it is list of the union type, and pydantic does not yet support that but they are planning on it